### ISSUE TITLE
Current Status

### ISSUE THREAD

**Issue by psigourney** [ericaddison/Hailing_Frequencies#27]
Finally able to get the actually chatty part working.   All the bells & whistles (use aliases, use the local user information, joining convos, muting people, etc) still need to be added, but at least it's got chat.

Just deployed it if you want to play around:

https://hailing-frequencies-2017.appspot.com

**Comment by ericaddison** [comment]
Hey man!

I'm back! I took a look at the deployed app ... looks pretty good! I'll take a poke at the back end tomorrow night to see about linking together the chat + email versions.

**Comment by mspagon** [comment]
Hope you had a wonderful trip!

I got the mobile app storing messages and retrieving messages (and other convo data) and I will sync up my data model to match yours.  I'm guessing you implemented convUsers, convMessages, Messages, and possibly some version of users?

I am spending a short bit learning fragments and various other android stuff in order to get navigation drawers and various menu options.  I will also start looking at how to use the NDB REST API and might have some questions.

**Comment by psigourney** [comment]
>  I'm guessing you implemented convUsers, convMessages, Messages, and possibly some version of users?

I'm not using Messages on ndb at all, only storing the actual chat messages to Firebase.  Using ndb for everything else though.

**Comment by ericaddison** [comment]
I'm thinking we will need to have the web and Android clients write messages to our main (NDB) server (through `POST /api/converstaions/<id>/messages/`) instead of Firebase, so our server can send out notifications to other comm options. I think our server should write a message to Firebase any time any message comes in (web, email, or text) in a unified way, so after that happens the web and android listeners will get their updates from Firebase.

**Comment by ericaddison** [comment]
this example:
https://github.com/GoogleCloudPlatform/python-docs-samples/tree/master/appengine/standard/firebase/firetactoe

has the front end send `POSTs` to the regular server, which then hits the Firebase DB, but also the front end listens for changes from Firebase.

**Comment by mspagon** [comment]
So this example uses both a regular server database (such as NDB) and Firebase DB? Is this a common thing? I PLAN on doing a NDB write immediately after doing a firebase write, but I just seemed that this isn't the usual case.

**Comment by mspagon** [comment]
I guess for now, @psigourney, we could just get the Firebase funcitonality working for the demo, just put in /TODO statements where we need put in the NDB writes so both databases can be consistent?  There has to be a better way, architecturally, I think having convos on both databases could introduce a ton of inconsistencies and bugs?

**Comment by ericaddison** [comment]
I don't know at all if this is typical!! ;) But I think the point is that we have some sort of server functionality beyond just simple DB access, which we do! Namely, we want different comm options to be able to talk with each other.

I'm mostly thinking of how to best integrate the different comm options (really just web and email right now). My thought is that all messages get processed in the same way, i.e. coming in through NDB, broadcasting through appropriate means (email and/or write to firebase DB). The NDB server will have to be able to write to the Firebase DB no matter what, since emails are processed through the NDB server and we want email messages to show up on the web chat ... but we also want web chat messages to be sent to email participants, so the NDB server will need to be aware of chat messages somehow, so I was thinking that funneling the web chat messages through the NDB server first, on the way to the Firebase DB, would help with that... Does that make sense?

**Comment by psigourney** [comment]
It should be easy to add a call to create a new message in the ndb datastore at the same time the Firebase entry is written.  

As for the ndb store updating Firebase when ndb receives a new entry... that'd take some more investigation.

**Comment by ericaddison** [comment]
Yeah... Here is what I understand from the tic-tac-toe example, for making a move in the game (analagous to posting a message):
- User makes a `POST` request to a URL on OUR server (not to firebase), see [this function](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/appengine/standard/firebase/firetactoe/firetactoe.py#L197)
- the `POST` handler calls a function to write to the firebase database , see [this function](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/master/appengine/standard/firebase/firetactoe/firetactoe.py#L98)
- that's it!

I think having a unified way to process different comm options is kind of important...

**Comment by mspagon** [comment]
If we have e-mail and SMS write directly to firebase it doesn't technically need to be written to NDB.  It would just be appended to the JSON?

**Comment by mspagon** [comment]
**_Edit: I re-read this thread and I see that we plan on simply making a POST request to our server and extend the API to write to Firebase.  Answered my question._**

So you are implying that it is relatively simple to extend the POST handlers to write to the firebase database and NDB database simultaneously?  I would think it is since the data is already there and writing to firebase is just another HTTP POST with a .json in the url?  (I would have done that while you were away but I don't really want to mess up the python code I don't understand.)

If this is true, we should definitely do it for posting messages to the NDB (POST /api/converstaions/<id>/messages/).  So the Android app can adhere to the REST methodology and just do HTTP POSTs for everything.

I ASSUMED that we were rushed and was manually writing (add convo and add message) to Firebase not using a REST call (since we don't have Firebase writes in our REST API).

How do you think we should proceed?

**Comment by ericaddison** [comment]
We want every comm option to be able to talk to every other comm option... I believe it should be possible, and probably without TOO much headache, to write to the Firebase DB from our Python server, so that such a thing could be done from a POST call...

On the Python side, a POST to `/api/conversations/<id>/messages/` (which contains the firebase token and json-ified message information) would end up calling [this](https://github.com/ericaddison/Hailing_Frequencies/blob/master/source/api/ConvMessages.py#L34) function, which would in turn call `broadcast_message()` (see [here](https://github.com/ericaddison/Hailing_Frequencies/blob/master/source/framework/communicate.py#L12)). Then there is a little method stub: [`send_to_web()`](https://github.com/ericaddison/Hailing_Frequencies/blob/master/source/framework/communicate.py#L34), which is where I think the call to Firebase DB should happen. I think filling in this one little method with a URL call to Firebase, the same way that you guys are doing on the front end but this time from python, I think that is all that is needed to make this work...

Then all incoming messages are handled in the same way, and broadcast out to all other users in the conversation no matter what comm option.

If you are up for it, try to fill in that [`send_to_web()`](https://github.com/ericaddison/Hailing_Frequencies/blob/master/source/framework/communicate.py#L34) Python method, which means putting in a call to Firebase to update the Firebase DB with the new conversation info. That update should then trigger a listener event back on the front ends. That fire tic-tac-toe sample has examples of how to call to the Firebase db from python.
If you have other pieces you can still work on, then I will give this a go tonight.

**Comment by mspagon** [comment]
Ok Great,
Just as an FYI:
I'm sure you already did some Firebase fooling around, but if you haven't:

Firebase is pretty simple for basic functionality.  (Android) To write, you just create a databaseReference and then write to a specific node (ConvMessages or ConvUsers, for example).  And to listen on the front end, you just create a listener for that node.  Both are like two lines of code.  And everything is automatic magic.

Each is a little different and i'm not familiar with anything besides Android.

Firebase for REST:
https://firebase.google.com/docs/database/rest/save-data

Firebase for WEB:
https://firebase.google.com/docs/database/web/read-and-write

Firebase for Android:
https://firebase.google.com/docs/database/android/read-and-write

**Comment by ericaddison** [comment]
nice! I think that the "Firebase for REST" will be what we need to call from the python side.

**Comment by mspagon** [comment]
Yes, you can use one of the Python helper libraries found here:

https://firebase.google.com/docs/database/rest/start

**Comment by ericaddison** [comment]
oh right ... NICE!

**Comment by psigourney** [comment]
Just modified the web so it POSTs to ConvMessages first, gets the Alias, then posts to Firebase using the alias.  Next thing would be to update the create_convMessage function to do the write to Firebase.

**Comment by ericaddison** [comment]
Right, but I think the right place to put in the Firebase call will be in the [`send_to_web()`](https://github.com/ericaddison/Hailing_Frequencies/blob/master/source/framework/communicate.py#L34) function. I'll look at that tonight.
